src/werkzeug/formparser.py

```diff
@@ -423,156 +411,322 @@ def validate_boundary(self, boundary):
             # the assert is skipped.
             self.fail("Boundary longer than buffer size")
 
-    def parse_lines(self, file, boundary, content_length, cap_at_buffer=True):
-        """Generate parts of
-        ``('begin_form', (headers, name))``
-        ``('begin_file', (headers, name, filename))``
-        ``('cont', bytestring)``
-        ``('end', None)``
+    class LineSplitter(object):
 
-        Always obeys the grammar
-        parts = ( begin_form cont* end |
-                  begin_file cont* end )*
-        """
-        next_part = b"--" + boundary
-        last_part = next_part + b"--"
-
-        iterator = chain(
-            make_line_iter(
-                file,
-                limit=content_length,
-                buffer_size=self.buffer_size,
-                cap_at_buffer=cap_at_buffer,
-            ),
-            _empty_string_iter,
-        )
+        """A stateful line splitter: call ``feed`` to push data in and receive
+        it split into lines.
 
-        terminator = self._find_terminator(iterator)
+        :param cap: Optional maximum length of a line; if this is given,
+                    lines will be truncated to meet it.
 
-        if terminator == last_part:
-            return
-        elif terminator != next_part:
-            self.fail("Expected boundary at start of multipart data")
-
-        while terminator != last_part:
-            headers = parse_multipart_headers(iterator)
+        ..versionadded:: 0.15
+        """
 
-            disposition = headers.get("content-disposition")
+        def __init__(self, cap=None):
+            self.buffer = b""
+            self.cap = cap
+
+        def _splitlines(self, pre, post):
+            # for most purposes, there is no difference between what we've
+            # just accepted and any other data. we accept them separately
+            # so we can tell if we're ending; see below.
+            buf = pre + post
+            rv = []
+            if not buf:
+                return rv, b""
+            lines = buf.splitlines(True)
+            iv = b""
+            for line in lines:
+                iv += line
+                while self.cap and len(iv) >= self.cap:
+                    rv.append(iv[: self.cap])
+                    iv = iv[self.cap :]
```

Comment:
In the case where self.cap is not None (or 0), we could do something more like this to avoid so many string copies:

```
for low_bound, high_bound in zip(range(0, len(iv), self.cap), range(self.cap, len(iv) + self.cap, self.cap)):
  rv.append(iv[low_bound:high_bound])
remainder = len(iv) % self.cap
iv = iv[-remainder:]
```